gdb: Expose CHERI SCRs via the GDB stub

bsdjhb opened the pull request:
This does not expose PCC and DDC in the new SCRs register group since those are already in the general purpose CHERI group.

bsdjhb committed:
gdb: Expose CHERI SCRs via the GDB stub

This does not expose PCC and DDC in the new SCRs register group since
those are already in the general purpose CHERI group.

bsdjhb commented on target/riscv/gdbstub.c:
```diff
@@ -228,6 +228,103 @@ static int riscv_gen_dynamic_csr_xml(CPUState *cs, int base_reg)
+    int index;
+    const char *name;
+    bool code;
+} scrs[] = {
```
This duplicates scr_info in cheri_op_helper.c a bit.  However, using all-lowercase names is more consistent with other registers exported via GDB.

bsdjhb commented on target/riscv/gdbstub.c:
```diff
@@ -228,6 +228,103 @@ static int riscv_gen_dynamic_csr_xml(CPUState *cs, int base_reg)
+    { .index = CheriSCR_MEPCC, .name = "mepcc", .code = true },
+};
+
+static inline cap_register_t *get_scr(CPUArchState *env, uint32_t index)
```
This is a clone of get_scr from cheri_op_helper.c.  Might be better to export that function instead?

jrtc27 commented on target/riscv/gdbstub.c:
```diff
@@ -228,6 +228,103 @@ static int riscv_gen_dynamic_csr_xml(CPUState *cs, int base_reg)
+    case CheriSCR_BSTDC: return &env->VSTDC;
+    case CheriSCR_BSScratchC: return &env->VSScratchC;
+    case CheriSCR_BSEPCC: return &env->VSEPCC;
+    default: assert(false);
```
I assume GDB will have already validated that the request from the wire was an advertised register and so this can't be triggered by asking for one of the encoding holes?

bsdjhb replied:
Correct.

jrtc27 commented on target/riscv/gdbstub.c:
```diff
@@ -228,6 +228,103 @@ static int riscv_gen_dynamic_csr_xml(CPUState *cs, int base_reg)
+    case CheriSCR_MScratchC: return &env->MScratchC;
+    case CheriSCR_MEPCC: return &env->MEPCC;
+
+    case CheriSCR_BSTCC: return &env->VSTCC;
```
Belong between S and M if including them, but I wouldn't bother, hypervisor bits will only really be done properly for the standardised version. Unless you're reusing the existing function of course, then it's fine for them to stay there.

bsdjhb committed:
Use upstream 'name' field for CSR names

This should fix the use of CSR names in the GDB stub.

bsdjhb committed:
riscv: Use all lower-case for SCR names in struct CPURISCVState

This matches the style for all other RISC-V registers including CSRs
and avoids a weird collision with STDC which zlib defines as an alias
of __STDC__.

bsdjhb committed:
riscv: Rename get_csr to riscv_get_csr and move to cpu.h

bsdjhb commented:
This also fixes CSRs exported by GDB to use their name instead of 'csrXXX'

jrtc27 reviewed (approved):
Just waiting for the Jenkins rerun to pass
